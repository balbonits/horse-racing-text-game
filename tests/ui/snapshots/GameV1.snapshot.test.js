/**
 * Snapshot Tests for GameV1 Integration UI Flows
 * 
 * Tests the complete game UI flows and interactions to ensure
 * consistent user experience across the v1.0 integration.
 */

const GameV1 = require('../../../src/GameV1');

// Mock blessed to prevent actual terminal output
jest.mock('blessed', () => ({
    screen: jest.fn(() => ({
        key: jest.fn(),
        append: jest.fn(),
        render: jest.fn(),
        destroy: jest.fn(),
        width: 80,
        height: 24
    })),
    box: jest.fn((config) => ({
        ...config,
        setContent: jest.fn(),
        destroy: jest.fn(),
        getContent: jest.fn(() => config.content || ''),
        focus: jest.fn()
    })),
    list: jest.fn((config) => ({
        ...config,
        on: jest.fn(),
        focus: jest.fn(),
        destroy: jest.fn()
    }))
}));

describe('GameV1 Integration UI Snapshots', () => {
    let game;
    let mockUI;

    beforeEach(() => {
        mockUI = {
            showLegalDisclaimer: jest.fn().mockResolvedValue({ accepted: true }),
            showOwnerCreation: jest.fn().mockResolvedValue({
                ownerName: 'Test Owner',
                stableName: 'Test Stables',
                philosophy: 'balanced'
            }),
            createScreen: jest.fn(),
            cleanup: jest.fn(),
            isInitialized: false,
            registerComponent: jest.fn(),
            getComponent: jest.fn(),
            destroyComponent: jest.fn()
        };
        
        game = new GameV1(mockUI);
    });

    afterEach(() => {
        if (game) {
            game.cleanup();
        }
    });

    describe('Game Status Snapshots', () => {
        test('should capture initial game status before initialization', () => {
            const status = game.getGameStatus();
            
            expect({
                gameStatus: status,
                isReady: game.isReady(),
                version: game.getVersion()
            }).toMatchSnapshot('initial-game-status');
        });

        test('should capture game status after initialization', async () => {
            await game.initialize();
            const status = game.getGameStatus();
            
            expect({
                gameStatus: status,
                isReady: game.isReady(),
                version: game.getVersion(),
                stableOwnerPresent: !!game.stableOwner,
                trainerDialogPresent: !!game.trainerDialog
            }).toMatchSnapshot('initialized-game-status');
        });

        test('should capture game status with current horse', async () => {
            await game.initialize();
            game.currentHorse = game.createRandomHorse();
            
            const status = game.getGameStatus();
            
            expect({
                gameStatus: status,
                currentHorse: {
                    name: game.currentHorse.name,
                    breed: game.currentHorse.breed,
                    stats: game.currentHorse.stats,
                    specialization: game.currentHorse.specialization
                }
            }).toMatchSnapshot('game-status-with-horse');
        });
    });

    describe('Horse Creation Snapshots', () => {
        beforeEach(async () => {
            await game.initialize();
        });

        test('should capture random horse creation results', () => {
            // Create multiple horses to show variation
            const horses = [];
            for (let i = 0; i < 5; i++) {
                horses.push(game.createRandomHorse());
            }
            
            const horseSnapshots = horses.map(horse => ({
                name: horse.name,
                breed: horse.breed,
                specialization: horse.specialization,
                racingStyle: horse.racingStyle,
                gender: horse.gender,
                stats: horse.stats
            }));
            
            expect(horseSnapshots).toMatchSnapshot('random-horse-creation-batch');
        });

        test('should capture horse serialization format', () => {
            const horse = game.createRandomHorse();
            const serialized = game.serializeHorse(horse);
            
            expect({
                originalHorse: {
                    name: horse.name,
                    stats: horse.stats,
                    breed: horse.breed,
                    specialization: horse.specialization,
                    racingStyle: horse.racingStyle,
                    gender: horse.gender
                },
                serializedFormat: serialized
            }).toMatchSnapshot('horse-serialization-format');
        });
    });

    describe('Trainer Integration Snapshots', () => {
        beforeEach(async () => {
            await game.initialize();
            game.currentHorse = game.createRandomHorse();
        });

        test('should capture complete trainer interaction flow', () => {
            const horse = game.currentHorse;
            const trainers = ['speed', 'stamina', 'power'];
            
            const trainerInteractions = trainers.map(specialty => {
                const trainer = game.getTrainer(specialty);
                const preMessage = game.getPreTrainingDialog(specialty, horse);
                
                // Simulate training
                const initialStat = horse.stats[specialty];
                const trainingResult = game.simulateTraining(horse, specialty);
                const postMessage = game.getPostTrainingDialog(specialty, trainingResult);
                
                return {
                    specialty,
                    trainer: trainer ? {
                        name: trainer.name,
                        specialty: trainer.specialty,
                        personality: trainer.personality.traits
                    } : null,
                    horse: {
                        name: horse.name,
                        initialStat: initialStat,
                        currentStat: horse.stats[specialty]
                    },
                    preMessage,
                    trainingResult,
                    postMessage
                };
            });
            
            expect(trainerInteractions).toMatchSnapshot('complete-trainer-interaction-flow');
        });

        test('should capture trainer relationship tracking', () => {
            const horse = game.currentHorse;
            const trainer = game.getTrainer('speed');
            
            const relationshipProgression = [];
            
            // Initial relationship
            relationshipProgression.push({
                sessions: 0,
                relationship: game.getTrainerRelationship(trainer.name, horse.name)
            });
            
            // Record several sessions
            for (let i = 1; i <= 5; i++) {
                game.recordTrainerSession(trainer.name, horse.name);
                relationshipProgression.push({
                    sessions: i,
                    relationship: game.getTrainerRelationship(trainer.name, horse.name)
                });
            }
            
            expect({
                trainer: trainer.name,
                horse: horse.name,
                progression: relationshipProgression
            }).toMatchSnapshot('trainer-relationship-tracking');
        });
    });

    describe('Name Generation Integration Snapshots', () => {
        beforeEach(async () => {
            await game.initialize();
        });

        test('should capture race generation with names', () => {
            const races = [
                game.generateRace('1200m', 'dirt'),
                game.generateRace('1600m', 'turf'),
                game.generateRace('2000m', 'dirt'),
                game.generateRace('2400m', 'turf')
            ];
            
            const raceSnapshots = races.map(race => ({
                name: race.name,
                trackName: race.trackName,
                distance: race.distance,
                surface: race.surface
            }));
            
            expect(raceSnapshots).toMatchSnapshot('generated-race-names');
        });

        test('should capture name uniqueness across session', () => {
            const horses = [];
            const races = [];
            
            for (let i = 0; i < 10; i++) {
                horses.push(game.createRandomHorse());
                races.push(game.generateRace('1600m', 'dirt'));
            }
            
            const horseNames = horses.map(h => h.name);
            const raceNames = races.map(r => r.name);
            const trackNames = races.map(r => r.trackName);
            
            expect({
                horseNames: {
                    names: horseNames,
                    uniqueCount: new Set(horseNames).size,
                    totalCount: horseNames.length
                },
                raceNames: {
                    names: raceNames,
                    uniqueCount: new Set(raceNames).size,
                    totalCount: raceNames.length
                },
                trackNames: {
                    names: trackNames,
                    uniqueCount: new Set(trackNames).size,
                    totalCount: trackNames.length
                }
            }).toMatchSnapshot('name-generation-uniqueness');
        });
    });

    describe('Save/Load Integration Snapshots', () => {
        beforeEach(async () => {
            await game.initialize();
            game.currentHorse = game.createRandomHorse();
        });

        test('should capture save data structure', () => {
            // Set up some game state
            game.currentTurn = 5;
            game.raceHistory = [
                { name: 'Test Race 1', result: 'won' },
                { name: 'Test Race 2', result: 'placed' }
            ];
            
            // Build trainer relationships
            const trainer = game.getTrainer('speed');
            for (let i = 0; i < 3; i++) {
                game.recordTrainerSession(trainer.name, game.currentHorse.name);
            }
            
            const saveData = game.save();
            
            expect({
                saveStructure: {
                    version: saveData.version,
                    timestamp: saveData.timestamp ? 'present' : 'missing',
                    stableOwner: saveData.stableOwner ? 'present' : 'missing',
                    currentHorse: saveData.currentHorse ? 'present' : 'missing',
                    nameGenerator: saveData.nameGenerator ? 'present' : 'missing',
                    trainerRelationships: saveData.trainerRelationships ? 'present' : 'missing',
                    gameState: saveData.gameState
                },
                stableOwnerData: saveData.stableOwner ? {
                    profile: saveData.stableOwner.profile,
                    financial: saveData.stableOwner.financial
                } : null,
                currentHorseData: saveData.currentHorse,
                trainerRelationshipsData: saveData.trainerRelationships
            }).toMatchSnapshot('save-data-structure');
        });

        test('should capture load restoration accuracy', () => {
            // Save original state
            const originalState = {
                stableOwnerName: game.stableOwner.profile.name,
                stableOwnerCash: game.stableOwner.financial.cash,
                horseName: game.currentHorse.name,
                horseStats: game.currentHorse.stats
            };
            
            const saveData = game.save();
            
            // Create new game and load
            const newGame = new GameV1(mockUI);
            const loadResult = newGame.load(saveData);
            
            const restoredState = {
                loadSuccessful: loadResult,
                stableOwnerName: newGame.stableOwner?.profile.name,
                stableOwnerCash: newGame.stableOwner?.financial.cash,
                horseName: newGame.currentHorse?.name,
                horseStats: newGame.currentHorse?.stats
            };
            
            expect({
                originalState,
                restoredState,
                dataMatches: {
                    stableOwnerName: originalState.stableOwnerName === restoredState.stableOwnerName,
                    stableOwnerCash: originalState.stableOwnerCash === restoredState.stableOwnerCash,
                    horseName: originalState.horseName === restoredState.horseName,
                    horseStatsMatch: JSON.stringify(originalState.horseStats) === JSON.stringify(restoredState.horseStats)
                }
            }).toMatchSnapshot('save-load-restoration-accuracy');
        });
    });

    describe('Error Handling Snapshots', () => {
        test('should capture initialization failure states', async () => {
            // Test with failing legal disclaimer
            mockUI.showLegalDisclaimer.mockResolvedValue({ accepted: false });
            
            const failedGame = new GameV1(mockUI);
            const result = await failedGame.initialize();
            
            expect({
                initializationResult: result,
                gameReady: failedGame.isReady(),
                stableOwnerPresent: !!failedGame.stableOwner
            }).toMatchSnapshot('initialization-failure-state');
        });

        test('should capture fallback behaviors', async () => {
            await game.initialize();
            
            const fallbackBehaviors = {
                invalidTraining: (() => {
                    try {
                        return game.simulateTraining(null, 'speed');
                    } catch (error) {
                        return { error: error.message };
                    }
                })(),
                invalidStatType: (() => {
                    try {
                        const horse = game.createRandomHorse();
                        return game.simulateTraining(horse, 'invalid');
                    } catch (error) {
                        return { error: error.message };
                    }
                })(),
                missingTrainer: game.getTrainer('invalid'),
                missingHorseDialog: game.getPreTrainingDialog('speed', null)
            };
            
            expect(fallbackBehaviors).toMatchSnapshot('error-fallback-behaviors');
        });
    });
});