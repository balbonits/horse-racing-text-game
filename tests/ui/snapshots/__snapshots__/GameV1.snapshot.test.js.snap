// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameV1 Integration UI Snapshots Error Handling Snapshots should capture fallback behaviors: error-fallback-behaviors 1`] = `
{
  "invalidStatType": {
    "error": "Invalid horse stats",
  },
  "invalidTraining": {
    "error": "Invalid training parameters",
  },
  "missingHorseDialog": "Let's get to work!",
  "missingTrainer": null,
}
`;

exports[`GameV1 Integration UI Snapshots Error Handling Snapshots should capture initialization failure states: initialization-failure-state 1`] = `
{
  "gameReady": false,
  "initializationResult": {
    "reason": "Legal disclaimer not accepted",
    "success": false,
  },
  "stableOwnerPresent": false,
}
`;

exports[`GameV1 Integration UI Snapshots Game Status Snapshots should capture game status after initialization: initialized-game-status 1`] = `
{
  "gameStatus": {
    "currentHorse": null,
    "currentTurn": 1,
    "initialized": true,
    "stableOwner": {
      "cash": 50000,
      "name": "Test Owner",
      "reputation": 0,
      "stableName": "Test Stables",
    },
    "version": "1.0",
  },
  "isReady": true,
  "stableOwnerPresent": true,
  "trainerDialogPresent": true,
  "version": "1.0",
}
`;

exports[`GameV1 Integration UI Snapshots Game Status Snapshots should capture game status with current horse: game-status-with-horse 1`] = `
{
  "currentHorse": {
    "breed": "thoroughbred",
    "name": "Royal Queen",
    "specialization": "miler",
    "stats": {
      "power": 27,
      "speed": 30,
      "stamina": 32,
    },
  },
  "gameStatus": {
    "currentHorse": {
      "breed": "thoroughbred",
      "name": "Royal Queen",
      "stats": {
        "power": 27,
        "speed": 30,
        "stamina": 32,
      },
    },
    "currentTurn": 1,
    "initialized": true,
    "stableOwner": {
      "cash": 50000,
      "name": "Test Owner",
      "reputation": 0,
      "stableName": "Test Stables",
    },
    "version": "1.0",
  },
}
`;

exports[`GameV1 Integration UI Snapshots Game Status Snapshots should capture initial game status before initialization: initial-game-status 1`] = `
{
  "gameStatus": {
    "currentHorse": null,
    "currentTurn": 1,
    "initialized": false,
    "stableOwner": null,
    "version": "1.0",
  },
  "isReady": false,
  "version": "1.0",
}
`;

exports[`GameV1 Integration UI Snapshots Horse Creation Snapshots should capture horse serialization format: horse-serialization-format 1`] = `
{
  "originalHorse": {
    "breed": "thoroughbred",
    "gender": "colt",
    "name": "Royal Queen",
    "racingStyle": "stalker",
    "specialization": "miler",
    "stats": {
      "power": 51,
      "speed": 51,
      "stamina": 40,
    },
  },
  "serializedFormat": {
    "bond": 0,
    "breed": "thoroughbred",
    "energy": undefined,
    "form": undefined,
    "gender": "colt",
    "name": "Royal Queen",
    "pedigree": null,
    "racingStyle": "stalker",
    "specialization": "miler",
    "stats": {
      "power": 51,
      "speed": 51,
      "stamina": 40,
    },
  },
}
`;

exports[`GameV1 Integration UI Snapshots Horse Creation Snapshots should capture random horse creation results: random-horse-creation-batch 1`] = `
[
  {
    "breed": "thoroughbred",
    "gender": "colt",
    "name": "Silver Legend",
    "racingStyle": "stalker",
    "specialization": "miler",
    "stats": {
      "power": 33,
      "speed": 33,
      "stamina": 56,
    },
  },
  {
    "breed": "thoroughbred",
    "gender": "colt",
    "name": "Thunder Spirit",
    "racingStyle": "stalker",
    "specialization": "miler",
    "stats": {
      "power": 56,
      "speed": 49,
      "stamina": 33,
    },
  },
  {
    "breed": "thoroughbred",
    "gender": "colt",
    "name": "Swift Queen",
    "racingStyle": "stalker",
    "specialization": "miler",
    "stats": {
      "power": 36,
      "speed": 35,
      "stamina": 32,
    },
  },
  {
    "breed": "thoroughbred",
    "gender": "colt",
    "name": "Gold Strike",
    "racingStyle": "stalker",
    "specialization": "miler",
    "stats": {
      "power": 46,
      "speed": 39,
      "stamina": 33,
    },
  },
  {
    "breed": "thoroughbred",
    "gender": "colt",
    "name": "Storm Dancer",
    "racingStyle": "stalker",
    "specialization": "miler",
    "stats": {
      "power": 48,
      "speed": 38,
      "stamina": 28,
    },
  },
]
`;

exports[`GameV1 Integration UI Snapshots Name Generation Integration Snapshots should capture name uniqueness across session: name-generation-uniqueness 1`] = `
{
  "horseNames": {
    "names": [
      "Storm King",
      "Lightning Legend",
      "Fire Strike",
      "Swift Flash",
      "Fire Flash",
      "Storm Spirit",
      "Storm Queen",
      "Gold Dancer",
      "Storm Spirit",
      "Lightning Flash",
    ],
    "totalCount": 10,
    "uniqueCount": 9,
  },
  "raceNames": {
    "names": [
      "Royal Prize",
      "Summer Trophy",
      "Moonrise Mile Classic",
      "Pine Grove Mile Feature",
      "Elmwood Mile Championship",
      "Golden Valley Mile Trophy",
      "Summer Feature",
      "Diamond Feature",
      "Spring Trial",
      "Sunset Mile Feature",
    ],
    "totalCount": 10,
    "uniqueCount": 10,
  },
  "trackNames": {
    "names": [
      "Golden Gardens",
      "Wind Grounds",
      "Pine Point Park",
      "Maple Valley Park",
      "Golden Valley Springs",
      "Southgate Valley",
      "Golden Park",
      "Pine Grove Speedway",
      "Silver Fields",
      "Silver Gardens",
    ],
    "totalCount": 10,
    "uniqueCount": 10,
  },
}
`;

exports[`GameV1 Integration UI Snapshots Name Generation Integration Snapshots should capture race generation with names: generated-race-names 1`] = `
[
  {
    "distance": "1200m",
    "name": "Autumn Special",
    "surface": "dirt",
    "trackName": "Thunder Park",
  },
  {
    "distance": "1600m",
    "name": "Winter Derby",
    "surface": "turf",
    "trackName": "Wind Meadows",
  },
  {
    "distance": "2000m",
    "name": "Winter Handicap",
    "surface": "dirt",
    "trackName": "Golden Park",
  },
  {
    "distance": "2400m",
    "name": "Autumn Festival",
    "surface": "turf",
    "trackName": "Storm Park",
  },
]
`;

exports[`GameV1 Integration UI Snapshots Save/Load Integration Snapshots should capture load restoration accuracy: save-load-restoration-accuracy 1`] = `
{
  "dataMatches": {
    "horseName": true,
    "horseStatsMatch": true,
    "stableOwnerCash": true,
    "stableOwnerName": true,
  },
  "originalState": {
    "horseName": "Fire Dancer",
    "horseStats": {
      "power": 28,
      "speed": 27,
      "stamina": 29,
    },
    "stableOwnerCash": 50000,
    "stableOwnerName": "Test Owner",
  },
  "restoredState": {
    "horseName": "Fire Dancer",
    "horseStats": {
      "power": 28,
      "speed": 27,
      "stamina": 29,
    },
    "loadSuccessful": true,
    "stableOwnerCash": 50000,
    "stableOwnerName": "Test Owner",
  },
}
`;

exports[`GameV1 Integration UI Snapshots Save/Load Integration Snapshots should capture save data structure: save-data-structure 1`] = `
{
  "currentHorseData": {
    "bond": 0,
    "breed": "thoroughbred",
    "energy": undefined,
    "form": undefined,
    "gender": "colt",
    "name": "Lightning King",
    "pedigree": null,
    "racingStyle": "stalker",
    "specialization": "miler",
    "stats": {
      "power": 50,
      "speed": 31,
      "stamina": 33,
    },
  },
  "saveStructure": {
    "currentHorse": "present",
    "gameState": {
      "currentTurn": 5,
      "isInitialized": true,
      "raceHistory": [
        {
          "name": "Test Race 1",
          "result": "won",
        },
        {
          "name": "Test Race 2",
          "result": "placed",
        },
      ],
    },
    "nameGenerator": "present",
    "stableOwner": "present",
    "timestamp": "present",
    "trainerRelationships": "present",
    "version": "1.0",
  },
  "stableOwnerData": {
    "financial": {
      "breedingIncome": 0,
      "cash": 50000,
      "facilityUpgradeCosts": 0,
      "monthlyOperatingCost": 5000,
      "netWorth": 50000,
      "prizeMoneyEarned": 0,
      "staffSalaries": 0,
      "totalEarnings": 0,
      "totalExpenses": 0,
    },
    "profile": {
      "bio": null,
      "experience": "Rookie",
      "founded": 2025-08-27T04:03:48.629Z,
      "name": "Test Owner",
      "philosophy": "balanced",
      "region": "Kentucky",
      "reputation": 0,
      "stableName": "Test Stables",
      "title": null,
    },
  },
  "trainerRelationshipsData": {
    "trainerRelationships": {
      "Coach Johnson-Lightning King": 3,
    },
  },
}
`;

exports[`GameV1 Integration UI Snapshots Trainer Integration Snapshots should capture complete trainer interaction flow: complete-trainer-interaction-flow 1`] = `
[
  {
    "horse": {
      "currentStat": 45,
      "initialStat": 43,
      "name": "Gold Runner",
    },
    "postMessage": "Steady progress! Every point of speed counts, and +2 is real improvement. Speed builds over time!

Good progress! Your horse is starting to show real speed potential in those sprints.",
    "preMessage": "Alright Gold Runner, time to see what you've got! We're going to work on your speed today - explosive starts and sustained acceleration!",
    "specialty": "speed",
    "trainer": {
      "name": "Coach Johnson",
      "personality": [
        "energetic",
        "direct",
        "results-focused",
      ],
      "specialty": "speed",
    },
    "trainingResult": {
      "improvement": 2,
      "newValue": 45,
      "previousValue": 43,
      "success": true,
      "trainer": "Coach Johnson",
    },
  },
  {
    "horse": {
      "currentStat": 54,
      "initialStat": 51,
      "name": "Gold Runner",
    },
    "postMessage": "Beautiful stamina building! +3 points - I can see you pushing through those fatigue barriers!

Nice progress! Those longer distances aren't looking as daunting anymore, are they?",
    "preMessage": "Welcome Gold Runner. Today we build endurance - the kind that wins races when others are finished. Steady and strong!",
    "specialty": "stamina",
    "trainer": {
      "name": "Coach Martinez",
      "personality": [
        "patient",
        "methodical",
        "endurance-focused",
      ],
      "specialty": "stamina",
    },
    "trainingResult": {
      "improvement": 3,
      "newValue": 54,
      "previousValue": 51,
      "success": true,
      "trainer": "Coach Martinez",
    },
  },
  {
    "horse": {
      "currentStat": 47,
      "initialStat": 45,
      "name": "Gold Runner",
    },
    "postMessage": "Solid strength gains! +2 power - real explosive ability is being forged here!

Good power development! I can see that extra thrust when your horse needs to accelerate.",
    "preMessage": "Gold Runner, time to build some real strength! Power training is tough, but that's what separates champions from the pack!",
    "specialty": "power",
    "trainer": {
      "name": "Coach Thompson",
      "personality": [
        "tough",
        "motivational",
        "fundamentals-focused",
      ],
      "specialty": "power",
    },
    "trainingResult": {
      "improvement": 2,
      "newValue": 47,
      "previousValue": 45,
      "success": true,
      "trainer": "Coach Thompson",
    },
  },
]
`;

exports[`GameV1 Integration UI Snapshots Trainer Integration Snapshots should capture trainer relationship tracking: trainer-relationship-tracking 1`] = `
{
  "horse": "Lightning Runner",
  "progression": [
    {
      "relationship": "new",
      "sessions": 0,
    },
    {
      "relationship": "familiar",
      "sessions": 1,
    },
    {
      "relationship": "familiar",
      "sessions": 2,
    },
    {
      "relationship": "familiar",
      "sessions": 3,
    },
    {
      "relationship": "familiar",
      "sessions": 4,
    },
    {
      "relationship": "experienced",
      "sessions": 5,
    },
  ],
  "trainer": "Coach Johnson",
}
`;
